object/encryptionkey.go

```diff
@@ -0,0 +1,270 @@
+	encoding.TextUnmarshaler
+	fmt.Stringer
+}
+
+var NoOpKey = EncryptionKey{
+	entropy: new([32]byte),
+	keyType: EncryptionKeyTypeBasic,
+}
+
+type EncryptionKeyType int
+
+const (
+	EncryptionKeyTypeBasic = EncryptionKeyType(iota + 1)
+	EncryptionKeyTypeSalted
+)
+
+// A EncryptionKey can encrypt and decrypt messages.
+type EncryptionKey struct {
+	entropy *[32]byte         `json:"-"`
+	keyType EncryptionKeyType `json:"-"`
+}
+
+// GenerateEncryptionKey returns a random encryption key.
+func GenerateEncryptionKey(t EncryptionKeyType) EncryptionKey {
+	key := EncryptionKey{
+		entropy: new([32]byte),
+		keyType: t,
+	}
+	frand.Read(key.entropy[:])
+	return key
+}
+
+func (k EncryptionKey) IsNoopKey() bool {
+	return bytes.Equal(k.entropy[:], NoOpKey.entropy[:])
+}
+
+func (k EncryptionKey) String() string {
+	if k.entropy == nil {
+		return ""
+	}
+	var prefix string
+	switch k.keyType {
+	case EncryptionKeyTypeBasic:
+		prefix = "key"
+	case EncryptionKeyTypeSalted:
+		prefix = "skey"
+	default:
+		return ""
+	}
+	return fmt.Sprintf("%s:%s", prefix, hex.EncodeToString(k.entropy[:]))
+}
+
+func (k EncryptionKey) Type() EncryptionKeyType {
+	return k.keyType
+}
+
+// MarshalBinary implements encoding.BinaryMarshaler.
+func (k EncryptionKey) MarshalBinary() ([]byte, error) {
+	var b [33]byte
+	switch k.keyType {
+	case EncryptionKeyTypeBasic:
+		b[0] = 1
+	case EncryptionKeyTypeSalted:
+		b[0] = 2
+	default:
+		return nil, ErrKeyType
+	}
+	copy(b[1:], k.entropy[:])
+	return b[:], nil
+}
+
+func (k *EncryptionKey) UnmarshalBinary(b []byte) error {
+	k.entropy = new([32]byte)
+	if len(b) != len(k.entropy)+1 {
+		return fmt.Errorf("wrong key length: expected %v, got %v", len(k.entropy)+1, len(b))
+	}
+	switch b[0] {
+	case 1:
+		k.keyType = EncryptionKeyTypeBasic
+	case 2:
+		k.keyType = EncryptionKeyTypeSalted
+	default:
+		return ErrKeyType
+	}
+	copy(k.entropy[:], b[1:])
+	return nil
+}
+
+// MarshalText implements the encoding.TextMarshaler interface.
+func (k EncryptionKey) MarshalText() ([]byte, error) {
+	return []byte(k.String()), nil
+}
+
+// UnmarshalText implements the encoding.TextUnmarshaler interface.
+func (k *EncryptionKey) UnmarshalText(b []byte) error {
+	splits := bytes.Split(b, []byte(":"))
+	if len(splits) != 2 {
+		return fmt.Errorf("expected the key to have the form prefix:entropy but had %v pieces", len(splits))
+	}
+	// prefix
+	switch string(splits[0]) {
+	case "key":
+		k.keyType = EncryptionKeyTypeBasic
+	case "skey":
+		k.keyType = EncryptionKeyTypeSalted
+	default:
+		return fmt.Errorf("invalid prefix for key: '%s'", splits[0])
+	}
+
+	// entropy
+	k.entropy = new([32]byte)
+	if n, err := hex.Decode(k.entropy[:], splits[1]); err != nil {
+		return err
+	} else if n != len(k.entropy) {
+		return fmt.Errorf("wrong key length: expected %v, got %v", len(k.entropy), n)
+	}
+	return nil
+}
+
+type EncryptionOptions struct {
+	Offset uint64
+	Key    *utils.UploadKey
+}
+
+var ErrKeyType = errors.New("invalid key type")
```

Comment:
`nit` can we move these up so we have `const` then `var` then `types` and so on

Reply:
(I can see why you define the type and then the methods though, but consts and vars should go up top imo)